core/src/main/java/com/linecorp/armeria/server/ExceptionReportingServerErrorHandler.java

```diff
@@ -0,0 +1,133 @@
+/*
+ * Copyright 2023 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.server;
+
+import java.time.Duration;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.LongAdder;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.linecorp.armeria.common.AggregatedHttpResponse;
+import com.linecorp.armeria.common.HttpResponse;
+import com.linecorp.armeria.common.HttpStatus;
+import com.linecorp.armeria.common.RequestHeaders;
+import com.linecorp.armeria.common.annotation.Nullable;
+import com.linecorp.armeria.common.util.TextFormatter;
+import com.linecorp.armeria.server.logging.LoggingService;
+
+/**
+ * A {@link ServerErrorHandler} that wraps another {@link ServerErrorHandler}
+ * to periodically report the exceptions that were not logged by decorators such as
+ * {@link LoggingService}.
+ */
+class ExceptionReportingServerErrorHandler implements ServerErrorHandler, ServerListener {
+
+    private static final Logger logger = LoggerFactory.getLogger(ExceptionReportingServerErrorHandler.class);
+
+    private final ServerErrorHandler delegate;
+    private final LongAdder counter;
+    private final Duration duration;
+    private boolean started;
+    @Nullable
+    private Throwable thrownException;
+    @Nullable
+    private ScheduledFuture<?> reportingTaskFuture;
+
+    ExceptionReportingServerErrorHandler(ServerErrorHandler serverErrorHandler, Duration duration) {
+        assert !duration.isNegative() && !duration.isZero();
+        delegate = serverErrorHandler;
+        this.duration = duration;
+        counter = new LongAdder();
+    }
+
+    /**
+     * If it's required to report exceptions, increase the counter and store first thrown exception.
+     */
+    @Nullable
+    @Override
+    public HttpResponse onServiceException(ServiceRequestContext ctx, Throwable cause) {
+        if (ctx.shouldReportUnloggedException()) {
+            if (thrownException == null) {
+                thrownException = cause;
+            }
+            counter.increment();
+        }
+        return delegate.onServiceException(ctx, cause);
```

Comment:
Some exceptions may be handled by a user-defined `ServerErrorHandler`.
The recovered exception is no longer logged by `LoggingService`.
We can't determine if the given `cause` is an unlogged exception.
The exceptions only handled by `DefaultServerErrorHandler` can be assumed unlogged exceptions.

So we may migrate the logic in `ExceptionReportingServerErrorHandler` to `DefaultServerErrorHandler` and record unlogged exceptions inside `DefaultServerErrorHandler`. It is necessary to make `DefaultServerErrorHandler` `class` and create it with a custom `unloggedExceptionReportInterval`.

Reply:
Okay, I missed that point. Please correct me if I misunderstood something. 
- Migrate `ExceptionReportingServerErrorHandler` logic into `DefaultServerErrorHandler` 
- Change `DefaultServerErrorHandler` type `enum` to `class` 
  - receive `unloggedExceptionReportInterval` as argument when being created 
  - log unhandled exceptions inside `DefaultServerErrorHandler`'s `onServiceException` method.

Reply:
Hmm, how about changing `ExceptionReportingServerErrorHandler` to call the delegate before recording then? Then a user-specified error handler has a chance to turn it off.

Reply:
@trustin , that seems to a great idea. So we can execute `delegate.onServiceException(ctx, cause)` first, and then count unlogged exception if needed. If user defined `ServerErrorHandler`s set `shouldReportUnloggedException` to false, we won't log the exceptions. 

@ikhoon , what do you think about this way ??

Reply:
>  execute delegate.onServiceException(ctx, cause) first, and then count unlogged exception if needed.

That means that if `delegate.onServiceException(ctx, cause)` returns null, do we regard `cause` as an unhandled exception?

Reply:
I was confused. 😅 I thought if a recovered response from an exception, the exception is not set to `RequestLog.responseCause()`. The exception is captured regardless of the error handler and set to `RequestLog.responseCause()`.
https://github.com/line/armeria/blob/ab206410ff1ce5602573b4e6009e0930e650b756/core/src/main/java/com/linecorp/armeria/server/HttpServerHandler.java#L376-L378

Sorry for the misdirection. `LoggingService` logs always when an error occurs and `ServerErrorHandler` is working separately. I think we don't need to call user-defined exception handlers. Because all exceptions are logged when `LoggingService` is added even if `ctx.shouldReportUnloggedException(false)` is called.

Reply:
@ikhoon thanks for the detailed explanation. If my understanding is correct, I think we can end this conversation without changing above code. Please correct me if I'm wrong.

Reply:
Right.